internal/app/prometheus-multi-tenant-proxy/server.go

```diff
@@ -6,41 +6,43 @@ import (
 	"net/http"
 	"net/http/httputil"
 	"net/url"
-	"os"
-	"sync"
 	"time"
 
-	"github.com/k8spin/prometheus-multi-tenant-proxy/internal/pkg"
 	"github.com/urfave/cli/v2"
 )
 
-var (
-	config     *pkg.Authn
-	configLock = new(sync.RWMutex)
-)
-
 // Serve serves
 func Serve(c *cli.Context) error {
 	prometheusServerURL, _ := url.Parse(c.String("prometheus-endpoint"))
 	serveAt := fmt.Sprintf(":%d", c.Int("port"))
 	authConfigLocation := c.String("auth-config")
 	reloadInterval := c.Int("reload-interval")
+	authType := c.String("auth-type")
+
+	var auth Auth
+	if authType == "basic" {
+		auth = NewBasicAuth(authConfigLocation)
+	} else if authType == "jwt" {
+		auth = NewJwtAuth(authConfigLocation)
+	} else {
+		log.Fatalf("auth-type must be one of: basic, jwt") // will exit
+	}
 
-	loadConfig(authConfigLocation)
-	ticker := time.NewTicker(time.Duration(reloadInterval) * time.Minute)
-	quit := make(chan struct{})
-	go func() {
-		for {
-			select {
-			case <-ticker.C:
-				loadConfig(authConfigLocation)
-				log.Printf("Reloaded config file %s", authConfigLocation)
-			case <-quit:
-				ticker.Stop()
-				return
+	if reloadInterval > 0 {
+		ticker := time.NewTicker(time.Duration(reloadInterval) * time.Minute)
+		quit := make(chan struct{})
+		go func() {
+			for {
+				select {
+				case <-ticker.C:
+					auth.Load()
```

Comment:
It could be a good idea maintaining a log message whenever the config is reloaded

Reply:
The logging is done in the `Load()` method. I moved it there because then we can only print something when there is actually a change. For basic, it logs every time, but for JWT it logs only when the JWKS has changed remotely.

Reply:
:O didn't noticed that. It's looking good to me :)